airflow/dag_processing/manager.py

```diff
@@ -709,14 +717,19 @@ def _add_callback_to_queue(self, request: CallbackRequest):
                 self.log.debug("Skipping already queued SlaCallbackRequest")
                 return
 
-            # not already queued, queue the file _at the back_, and add the request to the file's callbacks
+            # not already queued, queue the callback
+            # do NOT add the file of this SLA to self._file_path_queue. SLAs can arrive so rapidly that
+            # they keep adding to the file queue and never letting it drain. This in turn prevents us from
+            # ever rescanning the dags folder for changes to existing dags. We simply store the callback, and
+            # periodically, when self._file_path_queue is drained, we rescan and re-queue all DAG files.
+            # The SLAs will be picked up then. It means a delay in reacting to the SLAs (as controlled by the
+            # min_file_process_interval config) but stops SLAs from DoS'ing the queue.
             self.log.debug("Queuing SlaCallbackRequest for %s", request.dag_id)
             self._callback_to_execute[request.full_filepath].append(request)
-            if request.full_filepath not in self._file_path_queue:
-                self._file_path_queue.append(request.full_filepath)
+            Stats.incr("dag_processing.sla_callback_count")
```

Comment:
This change is the fix - we don't add to the queue. Note we _do_ still add the callback itself, so when the file is processed for some other reason (typically the periodic refresh of all dag files) the callback will be picked up then.